applications/OptimizationApplication/custom_utilities/container_variable_data_holder/container_variable_data_holder.cpp

```diff
@@ -0,0 +1,657 @@
+
+template<class TDataType>
+IndexType GetLocalSize(const IndexType DomainSize);
+
+template<>
+IndexType GetLocalSize<double>(const IndexType DomainSize)
+{
+    return 1;
+}
+
+template<>
+IndexType GetLocalSize<array_1d<double, 3>>(const IndexType DomainSize)
+{
+    return DomainSize;
+}
+
+void AssignValueToVector(
+    Vector& rOutput,
+    const IndexType StartIndex,
+    const IndexType ComponentIndex,
+    const double& rValue)
+{
+    rOutput[StartIndex] = rValue;
+}
+
+void AssignValueToVector(
+    Vector& rOutput,
+    const IndexType StartIndex,
+    const IndexType ComponentIndex,
+    const Array3D& rValue)
+{
+    rOutput[StartIndex + ComponentIndex] = rValue[ComponentIndex];
+}
+
+void AssignValueFromVector(
+    double& rOutput,
+    const IndexType ValueComponentIndex,
+    const IndexType VectoStartingIndex,
+    const Vector& rInput)
+{
+    rOutput = rInput[VectoStartingIndex];
+}
+
+void AssignValueFromVector(
+    Array3D& rOutput,
+    const IndexType ValueComponentIndex,
+    const IndexType VectoStartingIndex,
+    const Vector& rInput)
+{
+    rOutput[ValueComponentIndex] = rInput[VectoStartingIndex + ValueComponentIndex];
+}
+
+} // namespace ContainerVariableDataHolderHelperUtilities
+
+template <class TContainerType, class TContainerDataIO>
+typename ContainerVariableDataHolder<TContainerType, TContainerDataIO>::Pointer ContainerVariableDataHolder<TContainerType, TContainerDataIO>::Clone() const
+{
+    return std::make_shared<ContainerVariableDataHolder<TContainerType, TContainerDataIO>>(*this);
+}
+
+template <class TContainerType, class TContainerDataIO>
+typename ContainerVariableDataHolder<TContainerType, TContainerDataIO>::Pointer ContainerVariableDataHolder<TContainerType, TContainerDataIO>::CloneWithDataInitializedToZero() const
+{
+    std::shared_ptr<ContainerVariableDataHolder<TContainerType, TContainerDataIO>> result(new ContainerVariableDataHolder<TContainerType, TContainerDataIO>(*(this->mpModelPart)));
+    result->SetDataToZero(this->mDataDimension);
+    return result;
+}
+
+template <class TContainerType, class TContainerDataIO>
+template<class TDataType>
+void ContainerVariableDataHolder<TContainerType, TContainerDataIO>::ReadDataFromContainerVariable(const Variable<TDataType>& rVariable)
+{
+    KRATOS_TRY
+
+    KRATOS_ERROR_IF_NOT(this->GetModelPart().GetProcessInfo().Has(DOMAIN_SIZE))
+        << "DOMAIN_SIZE variable is not found in process info of model part "
+           "with name "
+        << this->GetModelPart().FullName() << ".\n";
+
+    const IndexType number_of_entities = this->GetContainer().size();
+    this->mDataDimension = ContainerVariableDataHolderHelperUtilities::GetLocalSize<TDataType>(this->GetModelPart().GetProcessInfo()[DOMAIN_SIZE]);
+
+    auto& r_data = this->mData;
+    auto& r_container = this->GetContainer();
+
+    r_data.resize(number_of_entities * this->mDataDimension);
+
+    IndexPartition<IndexType>(number_of_entities).for_each([&](const IndexType Index){
```

Comment:
it's usually good practice to explicitly state what you're capturing in a lambda instead of capturing everything by ref via `[&]`.
(for example, `number_of_entities` should be captured by value)